enclave/core/sgx/calls.c

```diff
@@ -775,6 +803,45 @@ oe_result_t oe_call_host_function_internal(
     /* Check the result */
     OE_CHECK(args->result);
 
+    return_args_ptr = (oe_call_function_return_args_t*)output_buffer;
+    /* Copy the marshaling struct from the host memory to avoid TOCTOU issues.
+     */
+    memcpy(
+        &return_args, return_args_ptr, sizeof(oe_call_function_return_args_t));
+
+    if (return_args.result == OE_OK)
+    {
+        /*
+         * Nonzero deepcopy_out_buffer and deepcopy_out_buffer_size fieldes
+         * indicate that there is deep-copied content that needs to be
+         * transmitted from the host. Also, ensuring that the content lies on
+         * host memory.
+         */
+        if (return_args.deepcopy_out_buffer &&
+            return_args.deepcopy_out_buffer_size &&
+            oe_is_outside_enclave(
+                return_args.deepcopy_out_buffer,
+                return_args.deepcopy_out_buffer_size))
+        {
+            void* enclave_buffer =
+                oe_malloc(return_args.deepcopy_out_buffer_size);
```

Comment:
Consider adding a function `oe_allocate_deep_copy_buffer` in oeedger8r/enclave.h:
https://github.com/openenclave/openenclave/blob/2f9661fb752f0e85c0964ed2a8e256dded2ca0ec/include/openenclave/edger8r/enclave.h#L130

That is consistent with the current practices. `oe_allocate_deep_copy_buffer` could in the future cache memory to avoid the ocall for memory allocation, or even use a buffer that is passed in by the host in ecall_contexte

Reply:
Currently the memory is freed via `oe_free` on the oeedger8r-generated code. Do you suggest updating the oeedger8r as well?

Reply:
Ideally yes. Edger8r tries to abstract host memory allocation/free functions for various reasons like potential optimization, well-defined interface etc.
But we could do that in another PR.

Reply:
Let's revisit this later. In this current ECALL design, the enclave edge routine allocates a host buffer and copies the in-enclave content over to the buffer. Then, the host edge routine is in charge of releasing the host buffer after unpacking the out parameters. In this case, we can't have the enclave to invoke the allocate & free pair by itself.